### ISSUE TITLE
Search and add queries that use Lucene specific optimizations

### ISSUE THREAD

**Issue by fulmicoton** [tantivy-search/search-benchmark-game#1]

**Comment by fulmicoton** [comment]
added a query that benefits heavily from the two step queries.

**Comment by usamec** [comment]
@fulmicoton any deeper description about what is exactly going on here? (link to some lucene documentation is enough).

**Comment by fulmicoton** [comment]
Right... So Lucene introduced two step queries a long time ago. The idea is that scorer can express that they have some notion of `.advanceWithFalsePositives()` that is cheap and a `checkIfReallyInDocset()` function that is expensive to compute.

This is typically the case for a phrase query. `.advanceWithFalsePositives()` is typically just an intersection of the terms. `checkIfReallyInDocset()`  requires to decode positionos and check if the terms align correctly somewhere in the document.

Now something interesting happens when running the query `"a b" AND c`.
The two step query system makes it possible to only decode positions for document that contain `C`.

**Comment by usamec** [comment]
Thanks, and is tantivy planning to add this?
